Framework/DataHandling/src/H5Util.cpp

```diff
@@ -395,6 +397,66 @@ template <typename NumT> std::vector<NumT> readArray1DCoerce(DataSet &dataset) {
   throw DataTypeIException();
 }
 
+/// Test if a group exists in an HDF5 file or parent group.
+bool groupExists(H5::H5Object &h5, const std::string &groupPath) {
+  bool status = true;
+  // Unfortunately, this is actually the approach recommended by the HDF Group.
+  try {
+    h5.openGroup(groupPath);
+  } catch (const H5::Exception &e) {
+    UNUSED_ARG(e);
+    status = false;
+  }
+  return status;
+}
+
+/// Test if an attribute is present on an HDF5 group or dataset and has a specific string value.
+bool keyHasValue(H5::H5Object &h5, const std::string &key, const std::string &value) {
+  bool status = true;
+  try {
+    Attribute attr = h5.openAttribute(key);
+    std::string value_;
+    attr.read(attr.getDataType(), value_);
+    if (value_ != value)
+      status = false;
+  } catch (const H5::Exception &e) {
+    UNUSED_ARG(e);
+    status = false;
+  }
+  return status;
+}
+
+void copyGroup(H5::H5Object &dest, const std::string &destGroupPath, H5::H5Object &src,
+               const std::string &srcGroupPath) {
+  // Source group must exist, and destination group must not exist.
+  if (!groupExists(src, srcGroupPath) || groupExists(dest, destGroupPath))
+    throw std::invalid_argument(std::string("H5Util::copyGroup: source group '") + srcGroupPath + "' must exist and " +
+                                "destination group '" + destGroupPath + "' must not exist.");
+
+  // TODO: check that source file must have at least read access and destination file must have write access.
+
+  // Note that in the HDF5 API:
+  //   C++ API support for these HDF5 methods does not yet exist.
+
+  // Create intermediate groups, if necessary
+  hid_t lcpl = H5Pcreate(H5P_LINK_CREATE);
+  if (H5Pset_create_intermediate_group(lcpl, 1) < 0)
+    throw std::runtime_error("H5Util::copyGroup: 'H5Pset_create_intermediate_group' error return.");
+
+  if (H5Ocopy(src.getId(), srcGroupPath.c_str(), dest.getId(), destGroupPath.c_str(), H5P_DEFAULT, lcpl) < 0)
+    throw std::runtime_error("H5Util::copyGroup: 'H5Ocopy' error return.");
```

Comment:
Will the error code appear in the logs despite not appearing in these exceptions?  I think that would be useful to have if not.

Reply:
This is a good point, and as far as I know the answer is _no_.  I'll rework these so there's at least some kind of error code in the exception-message cargo.

Reply:
I was wrong about that.  **I ran a test with this, and from the HDF5 "C" api, the error stack does appear in the `stderr` output stream.**  
Unfortunately, there's no simple way to add the string to the exception cargo.  The simplest thing the HDF5 "C" error API provides is a way to print the error stack to a "C"-style `FILE*`.    So, with that in mind, I'd prefer to leave this "as is".  **Let me know if you want me to pursue it further.**

Reply:
More explanation:  several of these new `H5Util` methods, use HDF5 C-api functions, that don't have a mapping in the current HDF5 C++ api.  That's why the "C" api error handling is what's relevant here.

Reply:
I see, so it just logs its own error and returns an error code, so it appears regardless.